python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/dygraph_sharding_optimizer.py

```diff
@@ -132,14 +128,35 @@ def _map_param_to_rank(self):
                 mapping[param.name] = rank
         return mapping
 
-    def _buid_inner_optimizer(self):
-        # we rely on the inner opt to determine whether a parameter is stop_gradient or not:
-        # create moment
-        # update related ops: clip, regular, opt
-        self._inner_optimizer = self._inner_optimizer_class(
-            parameters=self._rank2params[self._sharding_rank],
-            **self._inner_optimizer_kargs
-        )
+    def reduce_gradients(self, parameter_list, hcg):
+        # TODO merge grad / nrank with dp
+        logger.debug("sharding start gradients sync")
+        with framework.no_grad():
+            sharding_nrank = hcg.get_sharding_parallel_group().nranks
+            for param in parameter_list:
+                g_var = None
+                if param.trainable and (param._grad_ivar() is not None):
+                    g_var = param._grad_ivar()
+                if param.trainable and hasattr(param, "main_grad"):
+                    assert (
+                        param._grad_ivar() is None
+                    ), "param.grad should be None when using main_grad"
+                    g_var = param.main_grad
+                if g_var is not None:
+                    g_var.scale_(1.0 / sharding_nrank)
+                    param_rank = self._param2rank[param.name]
+                    paddle.distributed.all_reduce(
+                        g_var,
+                        group=hcg.get_sharding_parallel_group(),
+                        sync_op=True,
+                    )
+                    # TODO(pangengzheng): change to reduce operation when there is no diff in calculating global norm values in HybridParallelClipGrad compared to dp.
```

Comment:
There is a logic conflict between sharding stage 1 and sharding_pp_overlap which can be addressed in one of this two ways:
1. If use `all_reduce`, this part https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/distributed/fleet/meta_parallel/pp_utils/utils.py#L249 should also be updated, you can ref this pr https://github.com/PaddlePaddle/Paddle/pull/54872
2. If use `reduce`, then nothing else this comm method should be updated.